GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m

```diff
@@ -437,6 +443,75 @@ - (void)testDoWithFreshTokens_handleConcurrentRefresh {
   [self waitForExpectationsWithTimeout:1 handler:nil];
 }
 
+- (void)testAddScopes_success {
+  id signIn = OCMClassMock([GIDSignIn class]);
+  OCMStub([signIn sharedInstance]).andReturn(signIn);
+  [[signIn expect] addScopes:OCMOCK_ANY
+#if TARGET_OS_IOS || TARGET_OS_MACCATALYST
+    presentingViewController:OCMOCK_ANY
+#elif TARGET_OS_OSX
+            presentingWindow:OCMOCK_ANY
+#endif // TARGET_OS_IOS || TARGET_OS_MACCATALYST
+                  completion:OCMOCK_ANY];
+  
+  GIDGoogleUser *currentUser = [self googleUserWithAccessTokenExpiresIn:kAccessTokenExpiresIn
+                                                       idTokenExpiresIn:kIDTokenExpiresIn];
+  
+  OCMStub([signIn currentUser]).andReturn(currentUser);
+  
+#if TARGET_OS_IOS || TARGET_OS_MACCATALYST
+  UIViewController *presentingViewController = [[UIViewController alloc] init];
+  [currentUser addScopes:@[kNewScope]
+    presentingViewController:presentingViewController
+                  completion:nil];
+#elif TARGET_OS_OSX
+  NSWindow *presentingWindow = [[NSWindow alloc] init];
+  [currentUser addScopes:@[kNewScope]
+    presentingWindow:presentingWindow
+          completion:nil];
+#endif // TARGET_OS_IOS || TARGET_OS_MACCATALYST
+  
+  [signIn verify];
+}
+
+- (void)testAddScopes_failure_addScopesToPreviousAccount {
+  id signIn = OCMClassMock([GIDSignIn class]);
+  OCMStub([signIn sharedInstance]).andReturn(signIn);
+
+  GIDGoogleUser *currentUser = [self googleUserWithAccessTokenExpiresIn:kAccessTokenExpiresIn
+                                                       idTokenExpiresIn:kIDTokenExpiresIn];
+  
+  OCMStub([signIn currentUser]).andReturn(currentUser);
+  
+  GIDGoogleUser *previousUser = [self googleUserWithAccessTokenExpiresIn:kAccessTokenExpiresIn
+                                                        idTokenExpiresIn:kNewIDTokenExpiresIn];
+  
+  XCTestExpectation *expectation =
+      [self expectationWithDescription:@"Completion is called."];
+  
+#if TARGET_OS_IOS || TARGET_OS_MACCATALYST
+  UIViewController *presentingViewController = [[UIViewController alloc] init];
+  [previousUser addScopes:@[kNewScope]
+    presentingViewController:presentingViewController
+                  completion:^(GIDUserAuth *userAuth, NSError *error) {
+    [expectation fulfill];
+    XCTAssertNil(userAuth);
+    XCTAssertEqual(error.code, kGIDSignInErrorCodePreviousUser);
+  }];
+#elif TARGET_OS_OSX
+  NSWindow *presentingWindow = [[NSWindow alloc] init];
+  [previousUser addScopes:@[kNewScope]
+         presentingWindow:presentingWindow
+                  completion:^(GIDUserAuth *userAuth, NSError *error) {
+    [expectation fulfill];
+    XCTAssertNil(userAuth);
+    XCTAssertEqual(error.code, kGIDSignInErrorCodePreviousUser);
+  }];
+#endif // TARGET_OS_IOS || TARGET_OS_MACCATALYST
```

Comment:
I think the duplication of test code for iOS and macOS can be largely avoided here.

Reply:
Fair point. But if you inline `presentingViewController`/`presentingWindow`, `#if #elif #end` should be localized to one place. Up to you to decide if you want to do this here and other similar places.

Reply:
I think you should limit the `#if ... #endif` to the `presentingViewController` / `presentingWindow` line here as we do elsewhere in the codebase.  You're looking at dropping 30+ lines of repetitious code.

Reply:
Done.

Reply:
You missed one instance.